runners/jet/src/main/java/org/apache/beam/runners/jet/processors/TestStreamP.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.testing.TestStream;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.joda.time.Instant;
+
+/**
+ * Jet {@link com.hazelcast.jet.core.Processor} implementation for Beam's {@link TestStream}
+ * transform.
+ */
+public class TestStreamP extends AbstractProcessor {
+
+  private final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+  private final Traverser traverser;
+
+  @SuppressWarnings("unchecked")
+  private TestStreamP(List events, Coder outputCoder) {
+    traverser =
+        traverseIterable(events)
+            .map(
+                event -> {
+                  if (event instanceof SerializableWatermarkEvent) {
+                    long ts = ((SerializableWatermarkEvent) event).getTimestamp();
+                    if (ts == Long.MAX_VALUE) {
+                      // this is an element added by advanceWatermarkToInfinity(), we ignore it,
+                      // it's always at the end
+                      return null;
+                    }
+                    return new Watermark(ts);
+                  } else {
+                    assert event instanceof SerializableTimestampedValue;
+                    WindowedValue windowedValue =
+                        ((SerializableTimestampedValue) event).asWindowedValue();
+                    return Utils.encodeWindowedValue(windowedValue, outputCoder, baos);
+                  }
+                });
+  }
+
+  public static <T> ProcessorMetaSupplier supplier(
+      List<TestStream.Event<T>> events, Coder outputCoder) {
+    List<Object> serializableEvents = getSerializableEvents(events);
+    return ProcessorMetaSupplier.forceTotalParallelismOne(
+        ProcessorSupplier.of(() -> new TestStreamP(serializableEvents, outputCoder)));
+  }
+
+  private static <T> List<Object> getSerializableEvents(List<TestStream.Event<T>> events) {
+    return events.stream()
+        .flatMap(
+            e -> {
+              if (e instanceof TestStream.WatermarkEvent) {
+                return Stream.of(
+                    new SerializableWatermarkEvent(
+                        ((TestStream.WatermarkEvent<T>) e).getWatermark().getMillis()));
+              } else if (e instanceof TestStream.ElementEvent) {
+                return StreamSupport.stream(
+                        ((TestStream.ElementEvent<T>) e).getElements().spliterator(), false)
+                    .map(
+                        te -> new SerializableTimestampedValue<>(te.getValue(), te.getTimestamp()));
+              } else {
+                throw new UnsupportedOperationException(
+                    "Event type not supported in TestStream: " + e.getClass() + ", event: " + e);
+              }
+            })
+        .collect(toList());
+  }
+
+  @Override
+  public boolean complete() {
+    // todo: TestStream says it should cease emitting, but not stop after the items.
+    //   But I don't know how they end the job otherwise...
```

Comment:
TestStream should ensure that each of the events are processed independently. That can mean waiting for an event to be completely processed before sending the next.

Reply:
I understand what you are saying and indeed I've seen the need for such a behaviour myself, but I don't yet know how I could ensure such a behaviour when using Jet. Have taken a note and will look into before the next planned PR.